Fix aggregation function alias pushdown

brandboat opened the pull request:
Fixes #6189

Praveen2112 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushAggregationIntoTableScan.java:
```diff
@@ -225,11 +231,47 @@ private static AggregateFunction toAggregateFunction(Context context, Aggregatio
+
+    public static String getResolvedName(String name)
+    {
+        AggregationFunction variance =
```
Can we have generic way of resolving the alias name ?

brandboat replied:
@Praveen2112 Thanks for your comment ! I am not familiar with the part of SQL parsing. Currently I couldn't find out a way to get aggregation function alias name without specifically point out aggregation method/class which contain alias. 
I've thought about parsing all classes and methods under package `io.trino.operator.aggregation` and get those who contains alias name, but I'm not sure if this is a good idea, or maybe there is another better way.

Gentle ping @wendigo.

brandboat replied:
Finally figuring out another way to resolve alias name, could you take a look ? Many thanks ! 😃

Praveen2112 commented on core/trino-main/src/main/java/io/trino/operator/aggregation/AggregationFromAnnotationsParser.java:
```diff
@@ -127,20 +130,25 @@ private static AggregationHeader parseHeader(AnnotatedElement aggregationDefinit
     {
-        List<String> defaultNames = ImmutableList.<String>builder().add(aggregationAnnotation.value()).addAll(Arrays.asList(aggregationAnnotation.alias())).build();
+        Map<String, String> defaultNames = ImmutableMap.<String, String>builder()
+                .put(aggregationAnnotation.value(), aggregationAnnotation.value())
+                .putAll(Arrays.asList(aggregationAnnotation.alias()).stream().collect(
+                        Collectors.toMap(k -> k, v -> aggregationAnnotation.value())))
```
```suggestion
                .putAll(Stream.of(aggregationAnnotation.alias())
                        .collect(toImmutableMap(identity(), value -> aggregationAnnotation.value())))
```

Similarly in L162

brandboat replied:
We still need to keep this line.
``` java
.put(aggregationAnnotation.value(), aggregationAnnotation.value())
```
Otherwise we will miss parsing function with the resolved name.

Praveen2112 replied:
Yeah !! I accidentally copied that line

Praveen2112 commented on core/trino-main/src/main/java/io/trino/operator/aggregation/AggregationFromAnnotationsParser.java:
```diff
@@ -127,20 +130,25 @@ private static AggregationHeader parseHeader(AnnotatedElement aggregationDefinit
-        return getNames(toParse, aggregationAnnotation).stream()
-                .map(name ->
+        return getNames(toParse, aggregationAnnotation).entrySet().stream()
+                .map(nameEntry ->
```
```suggestion
                .map(entry ->
```
WDYT ?

brandboat replied:
Sure. 😃

Praveen2112 commented on core/trino-main/src/main/java/io/trino/operator/aggregation/AggregationHeader.java:
```diff
@@ -20,14 +20,21 @@
     private final boolean orderSensitive;
     private final boolean hidden;
 
     public AggregationHeader(String name, Optional<String> description, boolean decomposable, boolean orderSensitive, boolean hidden)
```
Can we modify `AggregationFromAnnotationsParser#parseHeader` to use the new constructor with the resolved name ?. And we can remove the old one

Praveen2112 commented on core/trino-main/src/main/java/io/trino/metadata/FunctionMetadata.java:
```diff
@@ -40,7 +41,7 @@ public FunctionMetadata(
+        this(FunctionId.toFunctionId(signature), signature, nullable, argumentDefinitions, hidden, deterministic, description, kind, false, signature.getName());
     }
 
     public FunctionMetadata(
```
Can we modify `ParametricScalar` and `SqlWindowFunction` to use the new constructor (with resolved name being passed as an parameter) ?

Praveen2112 reviewed (approved):
Thanks for working on this. % changes

dain commented:
I'd like to hear @martint's thoughts on this.  I expect this will be painful for the function changes, so maybe there is a better way to handle this.

martint commented on core/trino-main/src/main/java/io/trino/metadata/FunctionMetadata.java:
```diff
@@ -30,6 +30,7 @@
     private final String description;
     private final FunctionKind kind;
     private final boolean deprecated;
+    private final String resolvedName;
```
A better name for this would be "canonicalName"

brandboat replied:
@kasiafi Thanks for your comments, that helps a lot. 🙇

martint commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushAggregationIntoTableScan.java:
```diff
@@ -225,7 +225,8 @@ private static AggregateFunction toAggregateFunction(Context context, Aggregatio
         return new AggregateFunction(
-                signature.getName(),
+                metadata.getFunctionMetadata(aggregation.getResolvedFunction())
+                        .getResolvedName(),
```
Instead of carrying this name throughout the planning process, the function should be resolved to it's canonical signature during initial planning.

brandboat replied:
Thanks for your comments ! 😃 

I figure out a way, but not sure if this is appropriate. The way is to compare the origin name and the canonical name in Visitor.visitFunctionCall in ExpressionAnalyzer.java, if not equals, resolve the function to which has canonical name.

before
https://github.com/trinodb/trino/blob/cf68459a0d5d114d75bbb0897bca54654ae21a14/core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java#L1012
after
```java
function = metadata.resolveFunction(node.getName(), argumentTypes);

FunctionMetadata functionMetadata = metadata.getFunctionMetadata(function);
String originName = functionMetadata.getSignature().getName();
String canonicalName = metadata.getFunctionMetadata(function).getResolvedName();
if (!originName.equals(canonicalName)) {
    function = metadata.resolveFunction(QualifiedName.of(canonicalName), argumentTypes);
}
```

kasiafi commented:
I've scanned through the logs and can see 2 issues causing test failures.
1. Exceptions of the form
`Unsupported binding MODULUS(bigint, bigint):bigint for signature MODULUS(bigint,bigint):bigint`.
They come from the method `ParametricScalar.specialize()`. The immediate reason is that scalar function's `implementations` have a signature containing the non-canonical name, while `FunctionBinding` has a signature with the canonical name. As a result, they don't match. 

I tried to find a quick solution, but it seems that transition to consistently using the canonical name affects many sites. I think we could fix the issue with aggregations first, and deal with scalars as a follow-up.
@martint what is your opinion?

2. There's a bunch of planner tests, where actual plan doesn't match the expected plan. It happens because of function names being canonicalized. The expected plan has `rand()` but the actual has the canonical form `random()`etc. This should be easy to fix by canonicalizing the expressions in expected plans. However, we could consider alias-aware matching.

findepi commented:
> This should be easy to fix by canonicalizing the expressions in expected plans. However, we could consider alias-aware matching.

I think we should just updated the expected values in each affected test.

> Exceptions of the form
Unsupported binding MODULUS(bigint, bigint):bigint for signature MODULUS(bigint,bigint):bigint.
They come from the method ParametricScalar.specialize(). 

> I tried to find a quick solution, but it seems that transition to consistently using the canonical name affects many sites. 

What would it take to fix e.g. this particular case?

kasiafi commented:
> > Exceptions of the form
> > Unsupported binding MODULUS(bigint, bigint):bigint for signature MODULUS(bigint,bigint):bigint.
> > They come from the method ParametricScalar.specialize().
> 
> > I tried to find a quick solution, but it seems that transition to consistently using the canonical name affects many sites.
> 
> What would it take to fix e.g. this particular case?

Probably the quick fix would be to create `FunctionBinding` in `FunctionResolver.toFunctionBinding()` with a non-canonical name.
However, that would take us in the wrong direction if we intend to use the canonical names.

To go the other way, we need to update scalar implementations to carry the canonical name. This affects:
`ScalarFromAnnotationsParser.parseParametricScalar()` and `FunctionResolver.toFunctionBinding()` for consistency (this is the other source of BoundSignatures).
I applied these changes and still some issue remained with casts of rowtypes. I didn't investigate further.

kasiafi commented on core/trino-main/src/main/java/io/trino/metadata/FunctionMetadata.java:
```diff
@@ -22,60 +22,72 @@
+                        canonicalSignature.getReturnType(),
+                        canonicalSignature.getArgumentTypes(),
+                        canonicalSignature.isVariableArity())),
+                canonicalSignature, nullable, argumentDefinitions, hidden, deterministic, description, kind, deprecated, actualName);
```
Put arguments in separate lines.

kasiafi commented on core/trino-main/src/main/java/io/trino/operator/scalar/ParametricScalar.java:
```diff
@@ -47,7 +47,8 @@ public ParametricScalar(
             ParametricImplementationsGroup<ParametricScalarImplementation> implementations,
-            boolean deprecated)
+            boolean deprecated,
+            String actualName)
```
Let's not pass the actual name but take the name from the signature for now.

brandboat replied:
Do we need to keep argument `actualName` in ctor or just remove it ?

kasiafi replied:
We can remove it for now.

kasiafi commented on core/trino-main/src/test/java/io/trino/operator/TestAnnotationEngineForAggregates.java:
```diff
@@ -1157,6 +1160,73 @@ public void testPartiallyFixedTypeParameterInjectionAggregateFunctionParse()
+    @Test
+    public void testAggregateFunctionGetCanonicalName()
+    {
+        List<ParametricAggregation> aggOutputFuncList = parseFunctionDefinitions(AggregateOutputFunctionWithAlias.class);
```
Please do not use abbreviations.

kasiafi reviewed (commented):
Please add some tests to show how the change solves the original issue.

findepi commented:
> Please add some tests to show how the change solves the original issue.

I think that instead of, or additionally to, unit tests, we should have and end-to-end integration with connector APIs covered.
Fortunately, we don't need to mock anything, since the original problem was found in JDBC connectors aggregation pushdown and this is _why_ we're doing this.

@brandboat please search the code for occurrences of "https://github.com/trinodb/trino/issues/6189" (some TODO comments)

brandboat commented:
@findepi, Thanks for your comments ! 😃 
> I think that instead of, or additionally to, unit tests, we should have and end-to-end integration with connector APIs covered.
Fortunately, we don't need to mock anything, since the original problem was found in JDBC connectors aggregation pushdown and this is why we're doing this.
>
> @brandboat please search the code for occurrences of "#6189" (some TODO comments)

I've searched around TODO comments with #6189 and find out that we already have some tests that test aggregation pushdown in TestSqlServerIntegrationSmokeTest.java , TestPostgreSqlIntegrationSmokeTest.java. Does this means we already have end to end test ?

So we only need to remove alias name in the code for occurrences  of TODO #6189 comments, am I right ?

findepi commented:
> I've searched around TODO comments with #6189 and find out that we already have some tests that test aggregation pushdown in TestSqlServerIntegrationSmokeTest.java , TestPostgreSqlIntegrationSmokeTest.java. Does this means we already have end to end test ?

Yes, as long as these tests exercise the aliases -- which they do, e.g. here
https://github.com/trinodb/trino/blob/bf6b0d6b603cb027fe398666e1c52fff8f6518b1/plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java#L476

brandboat commented:
@kasiafi  Thanks for your reviews. I have addressed your comment by b00935d, f47ea0d, 1d50dfc, b424b6e.
Since we already have end to end test, b424b6e remove function alias names in `ImplementSqlServerStdev.java`, `ImplementSqlServerVariance.java`, `ImplementStddevSamp.java`, `ImplementVarianceSamp.java` that make sure only canonical names are accepted in jdbc connectors.

kasiafi commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/ImplementStddevSamp.java:
```diff
@@ -42,8 +42,7 @@
         return basicAggregation()
-                // TODO (https://github.com/trinodb/trino/issues/6189): remove stddev, an alias, from the list & simplify the pattern
-                .with(functionName().matching(name -> "stddev".equals(name) || "stddev_samp".equals(name)))
+                .with(functionName().matching(name -> "stddev".equals(name)))
```
This could be simplified to `.with(functionName().equalTo("stddev"))`

brandboat replied:
@kasiafi , I've addressed your comment in c12f028. Thanks for your review again.

kasiafi replied:
Thanks for updating! It looks good to me. @martint could you please take a look?

kasiafi commented:
LGTM.

findepi commented:
@brandboat can you please squash the commits?

brandboat commented:
> @brandboat can you please squash the commits?

copy that.

findepi commented on core/trino-main/src/main/java/io/trino/metadata/FunctionMetadata.java:
```diff
@@ -22,60 +22,91 @@
     private final String description;
     private final FunctionKind kind;
     private final boolean deprecated;
+    private final String actualName;
```
Please place this field just after `canonicalSignature` field, since they are closely related.

Please apply same orddering in constructor parameters, constructor assignments, and getters

brandboat replied:
copy that.

findepi commented on core/trino-main/src/main/java/io/trino/metadata/FunctionMetadata.java:
```diff
@@ -123,9 +154,14 @@ public boolean isDeprecated()
         return deprecated;
     }
 
+    public String getActualName()
```
Add a comment like
```
/**
 * Returns the name under which function is registered. Typically same as {@code getSignature().getName()}
 * unless this is an alias.
 */
```

brandboat replied:
copy that.

findepi commented on core/trino-main/src/main/java/io/trino/metadata/FunctionMetadata.java:
```diff
@@ -85,7 +116,7 @@ public FunctionId getFunctionId()
     public Signature getSignature()
     {
-        return signature;
+        return canonicalSignature;
```
If you do not rename the getter (which I am OK with), please keep the field name as it was.
Having different field and getter names creates confusion.

brandboat replied:
Indeed, this creates confusion. Rename methods getSignature will affect too many places, so I'll revert the naming.

findepi commented on core/trino-main/src/main/java/io/trino/operator/aggregation/AggregationFromAnnotationsParser.java:
```diff
@@ -127,20 +133,25 @@ private static AggregationHeader parseHeader(AnnotatedElement aggregationDefinit
     }
 
-    private static List<String> getNames(@Nullable AnnotatedElement outputFunction, AggregationFunction aggregationAnnotation)
+    private static Map<String, String> getNames(@Nullable AnnotatedElement outputFunction, AggregationFunction aggregationAnnotation)
```
The Map here is actually a pair of `(actual name, [actual name + aliases])` in disguise.
It's not incidental that all map values are always the same.
Ideally we should represent this with a dedicated value class (defined here, privately).
I am OK leaving this as is for now, but please add a comment explaining the returned Map structure, so that it doesn't surprise a reader in the future.

brandboat replied:
> Ideally we should represent this with a dedicated value class (defined here, privately).

Copy that.

findepi commented:
@brandboat thanks for quick update!

brandboat commented:
Thanks for the review ! 😃

kasiafi commented:
If you don't rename signature to canonicalSignature any more, please add an explanatory comment in `FunctionMetadata`.
There are 3 fields related to function name (`signature`, `actualName` and `functionId`), and their semantics isn't clear.
Please explain that `signature` is meant to contain canonical name, and that `functionId` is supposed to be consistent with the `actualName`.

brandboat commented:
> If you don't rename signature to canonicalSignature any more, please add an explanatory comment in FunctionMetadata.
There are 3 fields related to function name (signature, actualName and functionId), and their semantics isn't clear.
Please explain that signature is meant to contain canonical name, and that functionId is supposed to be consistent with the actualName.

@kasiafi , Thanks for the reminding, I'll add more comments to explain that. Might need some help since I am not a native speaker of English.

brandboat commented:
@kasiafi , I've addressed your comment in 8264e21. Thanks for your review again.

findepi commented:
Merged as 8db19d9c6b7ae87bf208e36c30303c30d39a3fb5, thanks!
I applied slightly edited version of the comments.

brandboat commented:
Thanks you so much for your patience and guidance!
